modules/proposals/engine/src/lib.rs

```diff
@@ -0,0 +1,348 @@
+            parameters,
+            title,
+            body,
+            proposer_id: proposer_id.clone(),
+            proposal_type,
+            status: ProposalStatus::Active,
+        };
+
+        // mutation
+        <Proposals<T>>::insert(new_proposal_id, new_proposal);
+        <ProposalCode>::insert(new_proposal_id, proposal_code);
+        ActiveProposalIds::mutate(|ids| ids.insert(new_proposal_id));
+        ProposalCount::put(next_proposal_count_value);
+
+        Self::deposit_event(RawEvent::ProposalCreated(proposer_id, new_proposal_id));
+
+        Ok(())
+    }
+}
+
+impl<T: Trait> Module<T> {
+    // Wrapper-function over system::block_number()
+    fn current_block() -> T::BlockNumber {
+        <system::Module<T>>::block_number()
+    }
+
+    // Executes approved proposal code
+    fn execute_proposal(proposal_id: u32) {
+        //let origin = system::RawOrigin::Root.into();
+        let proposal = Self::proposals(proposal_id);
+        let proposal_code = Self::proposal_codes(proposal_id);
+
+        let proposal_code_result =
+            T::ProposalCodeDecoder::decode_proposal(proposal.proposal_type, proposal_code);
+
+        let new_proposal_status = match proposal_code_result {
+            Ok(proposal_code) => {
+                if let Err(error) = proposal_code.execute() {
+                    ProposalStatus::Failed {
+                        error: error.as_bytes().to_vec(),
+                    }
+                } else {
+                    ProposalStatus::Executed
+                }
+            }
+            Err(error) => ProposalStatus::Failed {
+                error: error.as_bytes().to_vec(),
+            },
+        };
+
+        Self::update_proposal_status(proposal_id, new_proposal_status)
+    }
+
+    /// Voting results tally.
+    /// Returns proposals with changed status and tally results
+    fn tally() -> Vec<TallyResult<T::BlockNumber>> {
+        let mut results = Vec::new();
+        for &proposal_id in Self::active_proposal_ids().iter() {
+            let votes = Self::votes_by_proposal(proposal_id);
+            let proposal = Self::proposals(proposal_id);
+
+            if let Some(tally_result) = proposal.tally_results(
+                proposal_id,
+                votes,
+                T::TotalVotersCounter::total_voters_count(),
+                Self::current_block(),
+            ) {
+                results.push(tally_result);
+            }
+        }
+
+        results
+    }
+
+    /// Updates proposal status and removes proposal id from active id set.
+    fn update_proposal_status(proposal_id: u32, new_status: ProposalStatus) {
+        <Proposals<T>>::mutate(proposal_id, |p| p.status = new_status.clone());
+        ActiveProposalIds::mutate(|ids| ids.remove(&proposal_id));
+
+        Self::deposit_event(RawEvent::ProposalStatusUpdated(
+            proposal_id,
+            new_status.clone(),
+        ));
+
+        match new_status {
```

Comment:
Let's try to follow the idiom that `match` has no side-effects.

Reply:
Not found in Google. Explain, please.

Reply:
I also found nothing, but I was under the impression that the semantics implied by `match` was a computation, like `map`, where having side-effects would confuse the reader, but I see cases where even in the official website, there are side-effects (println)

https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html

So I guess I was wrong on this.